### PR TITLE
added discriminator for learner and facilitator

### PR DESCRIPTION
**Pull request by cherryyng**
## Notion ticket link
[Facilitator & Learner Relationship](https://www.notion.so/uwblueprintexecs/Facilitator-Learner-Relationship-20a31631fae246169c51ecb36c44b35c)

## Implementation description
* Added to base user model for learner and facilitator discriminators
* Added learner property to facilitators (list of strings)
* Added Facilitator property to learners (singular string)
* Created test in server.ts for saving to these user types

## Steps to test
1. Set up the .env with MG_DATABASE_NAME set to <name>_dev.
2. Run the server and verify that you can create and save documents with the new user types.
3. Check that the learners property for facilitators defaults to an empty array if not provided.
4. Ensure the facilitator property for learners is required and correctly assigned.

## What should reviewers focus on?
* Ensure the discriminators for the User model are correctly implemented.
* Verify the default value for the learners property.
* Confirm the facilitator property is required for learners and functions as expected.

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by jasondsouza35 — CHANGES_REQUESTED** [review]
I think it looks good overall cherry! I just have one comment which is in a highlighted code block.

**Review by achandrabalan — APPROVED** [review]
LGTM 🔥

File: backend/models/user.mgmodel.ts [hunk 1]
```diff
@@ -10,24 +10,46 @@ export interface User extends Document {
...
-    type: String,
-    required: true,
-  },
-  role: {
-    type: String,
-    required: true,
-    enum: ["Administrator", "Facilitator", "Learner"],
+const baseOptions = {
+  discriminatorKey: "type",
+  collection: "user",
```
Comment (carolynzhang18) [inline backend/models/user.mgmodel.ts:15]: I think we can remove this `collection` option.

My understanding is that `collection` specifies the name of the collection for the model? You'll notice we already had a collection `users` and that specifying this option spawns another collection `user`. No need for both, just `users` (the existing one) is fine.

Screenshot 2024-07-12 at 2 07 08 AM


FYI - I believe the default behaviour is that Mongoose comes up with a collection name by taking the lowercase plural of the first argument passed to `mongoose.model`, e.g. `mongoose.model<User>("User", UserSchema)` gives you `users`.

File: backend/models/user.mgmodel.ts [hunk 2]
```diff
@@ -10,24 +10,46 @@ export interface User extends Document {
...
-  authId: {
-    type: String,
-    required: true,
-  },
-  role: {
-    type: String,
-    required: true,
-    enum: ["Administrator", "Facilitator", "Learner"],
+const baseOptions = {
+  discriminatorKey: "type",
```
Comment (carolynzhang18) [inline backend/models/user.mgmodel.ts:14]: Hey actually... what if we have the `discriminatorKey` as `"role"` (the property that's already on User)?

By specifying a different discriminator key like `"type"`, we end up with the same thing showing up twice on facilitators and learners:
Screenshot 2024-07-12 at 2 10 54 AM

The catch is that you'll have to add an `Administrator` discriminator as well (or else creating an admin will cause an error) - should be ezpz, same as what you did for facilitator/learner discriminator, just with no additional properties (for now anyway) :D

**Review by carolynzhang18 — COMMENTED** [review]
Lookin really good!!! Just a few small fixes in `user.mgmodel.ts`, verify those changes, and then you can go ahead and remove the testing stuff in `server.ts` as well :D

File: backend/server.ts [hunk 3]
```diff
@@ -65,3 +68,38 @@ app.listen({ port: process.env.PORT || 8080 }, () => {
...
+      lastName: "Smith",
+      authId: "456",
+      role: "Learner",
+      facilitator: "123",
+    });
+    await learner.save();
+  } finally {
+    await mongoose.disconnect();
+  }
+}
```
[the remark this answers is not in the document]
Comment (cherryyng) [reply backend/server.ts:100]: Yes this will be removed after!

**Review by carolynzhang18 — APPROVED** [review]
image

**Pull request merged**